### PR TITLE
Update documentation for add-ons

### LINKED ISSUE #1372: Removed the Listener class but didn't updated ADDON.md
### Ruby version

3.3.0

### Code snippet

```
module RubyLsp
  module MyAddon
    class CodeLens < ::RubyLsp::Listener
    end
  end
end
```

### Description

I saw that when you bump version 0.14.0 the `Listener` class was removed but the Addons documentation was not updated, now i don't know how to create my own listeners with the new doc, because when I inherit form `::RubyLsp::Listener` the code breaks.

### Expected output

Expected a correct doc for the ADDON file

### PR DESCRIPTION
**Pull request by aryan-soni**
Resolves #1372 

### Motivation

Updating the add-on docs as per the changes in release `0.14.0`.

### Implementation

Updated `ADDONS.md`

### CODE REVIEW

File: ADDONS.md [hunk 1]
```diff
@@ -116,38 +117,35 @@ module RubyLsp
...
-    class Hover < ::RubyLsp::Listener
+    class Hover
       extend T::Sig
-      extend T::Generic
-
-      ResponseType = type_member { { fixed: T.nilable(::RubyLsp::Interface::Hover) } }
-
-      sig { override.returns(ResponseType) }
-      attr_reader :_response
+      include Requests::Support::Common
```
Comment (st0012) [inline ADDONS.md:134]:
```suggestion
      # the Requests::Support::Common module provides some helper methods you may find helpful
      include Requests::Support::Common
```

File: ADDONS.md [hunk 2]
```diff
@@ -116,38 +117,35 @@ module RubyLsp
...
           dispatcher: Prism::Dispatcher,
-        ).returns(T.nilable(Listener[T.nilable(Interface::Hover)]))
+        ).void
       end
-      def create_hover_listener(nesting, index, dispatcher)
+      def create_hover_listener(response_builder, nesting, index, dispatcher)
         # Use the listener factory methods to instantiate listeners with parameters sent by the LSP combined with any
         # pre-computed information in the addon. These factory methods are invoked on every request
-        Hover.new(@config, dispatcher)
+        Hover.new(client, response_builder, @config, dispatcher)
```
Comment (st0012) [inline ADDONS.md:129]: Let's also mention that it's possible to instantiate multiple listener classes in a comment?

File: ADDONS.md [hunk 3]
```diff
@@ -116,38 +117,35 @@ module RubyLsp
 
       sig do
         override.params(
+          response_builder: ResponseBuilders::Hover,
```
Comment (st0012) [inline ADDONS.md:120]: We need to mention that not all requests have a dedicated response builder. Some use `ResponseBuilders::CollectionResponseBuilder[ResponseDataType]` instead.

File: ADDONS.md [hunk 4]
```diff
@@ -116,38 +117,35 @@ module RubyLsp
...
-      ResponseType = type_member { { fixed: T.nilable(::RubyLsp::Interface::Hover) } }
-
-      sig { override.returns(ResponseType) }
-      attr_reader :_response
+      include Requests::Support::Common
 
       # Listeners are initialized with the Prism::Dispatcher. This object is used by the Ruby LSP to emit the events
       # when it finds nodes during AST analysis. Listeners must register which nodes they want to handle with the
       # dispatcher (see below).
+      # Also note that listeners are initialized with a `ResponseBuilders` object. The listener will push the
```
Comment (andyw8) [inline ADDONS.md:139]:
```suggestion
      # Listeners are initialized with a `ResponseBuilders` object. The listener will push the 
```

File: ADDONS.md [hunk 5]
```diff
@@ -159,10 +157,10 @@ module RubyLsp
       # define `on_constant_read_node_enter` to specify what this listener should do every time we find a constant
       sig { params(node: Prism::ConstantReadNode).void }
       def on_constant_read_node_enter(node)
-        # Certain helpers are made available to listeners to build LSP responses. The classes under `RubyLsp::Interface`
+        # Certain builders are made available to listeners to build LSP responses. The classes under `RubyLsp::ResponseBuilders`
         # are generally used to build responses and they match exactly what the specification requests.
```
Comment (andyw8) [inline ADDONS.md:161]:
```suggestion
        # are used to build responses conforming to the LSP Specification.
```

File: ADDONS.md [hunk 6]
```diff
@@ -114,40 +115,41 @@ module RubyLsp
         "Ruby LSP My Gem"
       end
 
+      # Not all requests have a dedicated response builder as shown in this example, some may use ResponseBuilders::CollectionResponseBuilder[ResponseDataType].
```
[the remark this answers is not in the document]
Comment (aryan-soni) [reply ADDONS.md]: Have raised this in issue #1413

**Review by vinistock — APPROVED** [review]
🚀

**Pull request merged**